pkg/filesystem/path/resolve.go

```diff
@@ -1,127 +1,60 @@
 	// Push the path without any leading slashes onto the stack, so
-	// that its components may be processed. Apply  them against the
+	// that its components may be processed. Apply them against the
 	// right directory.
-	componentWalker, err := scopeWalker.OnScope(absolute)
+	componentWalker, remainder, err := parser.ParseScope(scopeWalker)
 	if err != nil {
 		return err
 	}
-	rs.stack = append(rs.stack, path)
+	if remainder != nil {
+		rs.stack = append(rs.stack, remainder)
+	}
 	rs.componentWalker = componentWalker
 	return nil
 }
 
 // Pop a single filename from the stack of paths that need to be
 // processed. This filename is the first one that should be processed.
-func (rs *resolverState) pop() string {
+func (rs *resolverState) pop() (GotDirectoryOrSymlink, error) {
 	p := &rs.stack[len(rs.stack)-1]
-	slash := strings.IndexByte(*p, '/')
-	if slash == -1 {
-		// Path no longer contains a slash. Consume it entirely.
-		name := *p
+	node, remainder, err := (*p).ParseFirstComponent(rs.componentWalker, len(rs.stack) > 1)
+	if err != nil {
+		return nil, err
+	}
+	if remainder == nil {
 		rs.stack = rs.stack[:len(rs.stack)-1]
-		return name
+	} else {
+		rs.stack[len(rs.stack)-1] = remainder
 	}
-
-	// Consume the next component and as many slashes as possible.
-	name := (*p)[:slash]
-	*p = stripOneOrMoreSlashes((*p)[slash:])
-	return name
+	return node, err
 }
 
 func (rs *resolverState) resolve() error {
 	for len(rs.stack) > 0 {
-		switch name := rs.pop(); name {
-		case "", ".":
-			// An explicit "." entry, or an empty component.
-			// Empty components can occur if paths end with
-			// one or more slashes. Treat "foo/" identical
-			// to "foo/."
-		case "..":
-			// Traverse to the parent directory.
-			componentWalker, err := rs.componentWalker.OnUp()
-			if err != nil {
-				return err
-			}
-			rs.componentWalker = componentWalker
-		default:
-			if len(rs.stack) > 0 {
-				// A filename that was followed by a
-				// slash, or we are symlink expanding
-				// one or more paths that are followed
-				// by a slash. This component must yield
-				// a directory or symlink.
-				r, err := rs.componentWalker.OnDirectory(Component{
-					name: name,
-				})
-				if err != nil {
-					return err
-				}
-				switch rv := r.(type) {
-				case GotDirectory:
-					rs.componentWalker = rv.Child
-				case GotSymlink:
-					if err := rs.push(rv.Parent, rv.Target); err != nil {
-						return err
-					}
-				default:
-					panic("Missing result")
-				}
-			} else {
-				// This component may be any kind of file.
-				r, err := rs.componentWalker.OnTerminal(Component{
-					name: name,
-				})
-				if err != nil || r == nil {
-					// Path resolution ended with
-					// any file other than a symlink.
-					return err
-				}
-				// Observed a symlink at the end of a
-				// path. We should continue to run.
-				if err := rs.push(r.Parent, r.Target); err != nil {
+		r, err := rs.pop()
+		if err != nil {
+			return err
+		}
+		if r != nil {
```

Comment:
Apparently it is permitted to use `case nil:` in `switch x.(type)`. Maybe use that, so that we can drop this if-statement?